src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php

```diff
@@ -0,0 +1,218 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Adapter\PhpAmqp;
+
+/**
+ * An AMQP connection.
+ *
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class Connection
+{
+    private $connectionCredentials;
+    private $exchangeConfiguration;
+    private $queueConfiguration;
+    private $debug;
+    private $amqpFactory;
+
+    /**
+     * @var \AMQPChannel|null
+     */
+    private $amqpChannel;
+
+    /**
+     * @var \AMQPExchange|null
+     */
+    private $amqpExchange;
+
+    /**
+     * @var \AMQPQueue|null
+     */
+    private $amqpQueue;
+
+    public function __construct(array $connectionCredentials, array $exchangeConfiguration, array $queueConfiguration, bool $debug = false, AmqpFactory $amqpFactory = null)
+    {
+        $this->connectionCredentials = $connectionCredentials;
+        $this->debug = $debug;
+        $this->exchangeConfiguration = $exchangeConfiguration;
+        $this->queueConfiguration = $queueConfiguration;
+        $this->amqpFactory = $amqpFactory ?: new AmqpFactory();
+    }
+
+    public static function fromDsn(string $dsn, array $options = array(), bool $debug = false, AmqpFactory $amqpFactory = null): self
```

Comment:
What about a Trait or some reusable method for the parsing stuff?

Reply:
I'd use the same point used in the previous comment: [we don't know yet so let's not expose too much](https://github.com/symfony/symfony/pull/26632#discussion_r178805235). Or said another way: what is public (i.e. to be able to be shared/re-use) is supersensible to BC-break and adds up to our BC responsibility. Adding such exposure for an hypothetic use case is too risky IMHO.